src/Storages/PostgreSQL/PostgreSQLConnection.h

```diff
@@ -7,44 +7,73 @@
 #if USE_LIBPQXX
 #include <pqxx/pqxx> // Y_IGNORE
 #include <Core/Types.h>
+#include <Common/ConcurrentBoundedQueue.h>
 
 
 namespace DB
 {
 
-/// Tiny connection class to make it more convenient to use.
-/// Connection is not made until actually used.
 class PostgreSQLConnection
 {
-public:
-    using ConnectionPtr = std::shared_ptr<pqxx::connection>;
-
-    PostgreSQLConnection(std::string dbname, std::string host, UInt16 port, std::string user, std::string password);
 
-    PostgreSQLConnection(const PostgreSQLConnection & other);
+using ConnectionPtr = std::shared_ptr<pqxx::connection>;
 
-    PostgreSQLConnection operator =(const PostgreSQLConnection &) = delete;
+public:
+    PostgreSQLConnection(
+        const String & connection_str_,
+        const String & address_);
 
-    bool tryConnect();
+    PostgreSQLConnection(const PostgreSQLConnection & other) = delete;
 
-    ConnectionPtr conn();
+    ConnectionPtr get();
 
-    const std::string & getAddress() { return address; }
+    ConnectionPtr tryGet();
 
-    std::string & conn_str() { return connection_str; }
+    bool isConnected() { return tryConnectIfNeeded(); }
 
 private:
-    void connect();
+    void connectIfNeeded();
+
+    bool tryConnectIfNeeded();
 
-    static std::string formatConnectionString(
-        std::string dbname, std::string host, UInt16 port, std::string user, std::string password);
+    const std::string & getAddress() { return address; }
 
     ConnectionPtr connection;
     std::string connection_str, address;
 };
 
 using PostgreSQLConnectionPtr = std::shared_ptr<PostgreSQLConnection>;
 
+
+class PostgreSQLConnectionHolder
+{
+
+using Pool = ConcurrentBoundedQueue<PostgreSQLConnectionPtr>;
+using PoolPtr = std::shared_ptr<Pool>;
+
+public:
+    PostgreSQLConnectionHolder(PostgreSQLConnectionPtr connection_, PoolPtr pool_)
+        : connection(std::move(connection_))
+        , pool(std::move(pool_))
+    {
+    }
+
+    PostgreSQLConnectionHolder(const PostgreSQLConnectionHolder & other) = delete;
+
+    ~PostgreSQLConnectionHolder() { pool->tryPush(connection); }
+
+    pqxx::connection & conn() const { return *connection->get(); }
+
+    bool isConnected() { return connection->isConnected(); }
+
+private:
+    PostgreSQLConnectionPtr connection;
+    PoolPtr pool;
```

Comment:
An owning pointer to another object's member is somewhat weird. This means that the pool and its queue can have different lifetimes. Do you need a shared pointer here? Maybe lifetime of connections can't be more than lifetime of pool, and you can use a weak_ptr here? I think if the pool is destroyed when the storage is destroyed, it should work, because the streams that use the connections shouldn't live longer than storage.

Reply:
Or even not a weak_ptr but just a reference -- you don't need to handle the case when the pool is destroyed earlier, because it would be an error.